pageserver/src/tenant.rs

```diff
@@ -2665,17 +2642,15 @@ impl Tenant {
             src_timeline.pg_version,
         );
 
-        let new_timeline = {
-            let mut timelines = self.timelines.lock().unwrap();
-            self.prepare_timeline(
-                dst_id,
-                &metadata,
-                timeline_uninit_mark,
-                false,
-                Some(Arc::clone(src_timeline)),
-            )?
-            .initialize_with_lock(ctx, &mut timelines, true)?
-        };
+        let uninitialized_timeline = self.prepare_new_timeline(
+            dst_id,
+            &metadata,
+            timeline_uninit_mark,
+            start_lsn + 1,
+            Some(Arc::clone(src_timeline)),
+        )?;
```

Comment:
The `start_lsn + 1` is confusing to me.

I think understand why we do it: to set the right begin lsn for the open layer.

But the other callers of `prepare_timeline` don't use it.

@hlinnaka please explain why it's needed here and not with the other callers.

Reply:
It was a faithful conversion from the old behavior. This PR replaces the `prepare_timeline` function with `prepare_new_timeline` (or renames it, if you will). There were three callers, with two different ways of calling it:

1. prepare_timeline(init_layers == true)
2. prepare_timeline(init_layers == false), followed by initialize_with_lock(load_layer_map=true)

In the first case, prepare_timeline initialized `next_open_layer_at` with this:
```
                    new_timeline.layers.write().unwrap().next_open_layer_at =
                        Some(new_timeline.initdb_lsn);
```

In the second case, load_layer_map() initialized it like this:
```
        layers.next_open_layer_at = Some(Lsn(disk_consistent_lsn.0) + 1);
```

Now, whether that _ever_ made sense is another question :-). When branching, the +1 is necessary, because otherwise we would miss the last record on the parent branch when traversing to the parent. It might make sense to always do the +1, though. Or change the way `get_reconstruct_data` traverses to the parent. Or when creating a new branch, maybe the caller should already pass start_lsn +1. This isn't very well-thought out or documented...

Reply:
> This isn't very well-thought out or documented...

Yeah, I figured. Thanks for the explanation, makes sense.

I'll not aim for improving it in this PR, just posting my audit of your faithful conversion.

Before:

|                      |                      | before ingest                |                     | after ingest                 |                    | comment                                                                                                                             |
|----------------------|----------------------|------------------------------|---------------------|------------------------------|--------------------|-------------------------------------------------------------------------------------------------------------------------------------|
|                      | initialize_with_lock | metadata disk consistent lsn | next open layer     | metadata disk consistent lsn | next open layer    |                                                                                                                                     |
| import basebackup    | load_layers=false    | base_lsn                     | base_lsn            | base_lsn + X                 | base_lsn + X + 1   | We get the X+1 from the<br>checkpoint at end of ingest                                                                              |
| bootstrap            | load_layers=false    | initdb_lsn                   | initdb_lsn          | initdb_lsn + X               | initdb_lsn + X + 1 | We get the X+1 from the<br>checkpoint at end of ingest                                                                              |
| branching            | no                   | branchpoint_lsn              | branchpoint_lsn + 1 | there is no ingest           | there is no ingest | We get the +1 because we call<br>prepare_timeline with branchpoint_lsn+1<br>NB: branchpoint_lsn is called<br> start_lsn in the code |
| create_test_timeline | load_layers=true     | the `initdb_lsn` arg         | `initdb_lsn + 1`    | there is no ingest           | there is no ingest | We get the +1 through the load_layers=true                                                                                          |

After

|                      | before ingest                |                     | after ingest                 |                    | comment                                                                                                                             |
|----------------------|------------------------------|---------------------|------------------------------|--------------------|-------------------------------------------------------------------------------------------------------------------------------------|
|                      | metadata disk consistent lsn | next open layer     | metadata disk consistent lsn | next open layer    |                                                                                                                                     |
| import basebackup    | base_lsn                     | base_lsn            | base_lsn + X                 | base_lsn + X + 1   | We get the X+1 from the<br>checkpoint at end of ingest                                                                              |
| bootstrap            | initdb_lsn                   | initdb_lsn          | initdb_lsn + X               | initdb_lsn + X + 1 | We get the X+1 from the<br>checkpoint at end of ingest                                                                              |
| branching            | branchpoint_lsn              | branchpoint_lsn + 1 | there is no ingest           | there is no ingest | We get the +1 because we call<br>prepare_timeline with branchpoint_lsn+1<br>NB: branchpoint_lsn is called<br> start_lsn in the code |
| create_test_timeline | the `initdb_lsn` arg         | `initdb_lsn` arg    | there is no ingest           | there is no ingest | **next_open_layer doesn't start at +1!?**                                                                                           |


So, the third three are the same but `create_test_timeline` changed behavior with your refactoring.

Before, it would have `next_open_layer = initdb_lsn + 1`.

After, it has `next_open_layer = initdb_lsn`.

Reply:
Ok, good catch, I didn't think about the `create_test_timeline` case.

That seems good to me. create_test_timeline feels similar to the bootstrapping or import_basebackup case, so it makes sense to do the same thing.

Reply:
In the meantime I, together with @LizardWizzard , came to a conclusion that resulted in https://github.com/neondatabase/neon/pull/4451

Please take a look there, it's meant as a preliminary for this PR. Will rebase this PR on top. The rebase will also allow us to get rid of the cfg(test) set_disk_consistent_lsn function that you added in this PR.